File: bridge/resources/android/LemonfiberFunctions.kt
package app.lemonfiber.native

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.util.Log
import android.view.WindowManager
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction

/**
 * The Android half of lemonfiber's own native expansion.
 *
 * Namespace: `Lemonfiber.*`
 *
 * Everything here is a thin shim over one window flag. The decision about *when*
 * that flag belongs on lives in [CaptureRule], which imports nothing from the
 * framework and is unit-tested; this file is what cannot be tested off a device
 * and is deliberately kept down to the calls that need one.
 *
 * `FLAG_SECURE` is Android's whole answer to both requirements. One flag blanks
 * the recents-screen thumbnail and refuses screenshots and screen recording
 * together. There is no finer-grained control: a view cannot opt out of a
 * capture on its own, so the flag goes on the window or the protection does not
 * exist.
 *
 * **It is applied on `onPause`, not held forever.** Holding it permanently would
 * be simpler and is what most examples do, and it would also refuse every
 * deliberate screenshot for the life of the app — including the one an operator
 * takes to send to somebody who can help. The recents snapshot is taken after
 * `onPause`, so setting it there is in time for the only capture the
 * task-switcher rule is about. A guarded screen sets it in the foreground too,
 * and keeps it.
 */
public object LemonfiberFunctions {
    private const val TAG = "Lemonfiber"

    /**
     * What the app is showing and whether anybody is looking at it.
     *
     * Process-wide because the window is. A companion holds one activity, and
     * two screens cannot disagree about whether the window is protected — so
     * this is one value rather than a per-screen flag that the last writer wins.
     */
    @Volatile
    private var rule: CaptureRule = CaptureRule.LAUNCHED

    /**
     * Start watching the app move in and out of the foreground.
     *
     * Called once from `installLemonfiber`, which is the symbol the builder
     * calls. Without it the window is never protected, because nothing would
     * notice the app leaving — and the failure would be invisible: every screen
     * would look right, and the task switcher would quietly hold the last frame.
     */
    @JvmStatic
    public fun install(application: Application) {
        application.registerActivityLifecycleCallbacks(
            object : Application.ActivityLifecycleCallbacks {
                override fun onActivityPaused(activity: Activity) {
                    rule = rule.backgrounded()
                    apply(activity)
                }

                override fun onActivityResumed(activity: Activity) {
                    rule = rule.foregrounded()
                    apply(activity)
                }

                override fun onActivityCreated(
                    activity: Activity,
                    state: Bundle?,
                ) = Unit

                override fun onActivityStarted(activity: Activity) = Unit

                override fun onActivityStopped(activity: Activity) = Unit

                override fun onActivitySaveInstanceState(
                    activity: Activity,
                    out: Bundle,
                ) = Unit

                override fun onActivityDestroyed(activity: Activity) = Unit
            },
        )
    }

    /**
     * Put the window into whatever state the rule now calls for.
     *
     * The flag is set and cleared rather than toggled, so a state that is
     * already correct costs nothing and a state that drifted is corrected. A
     * toggle would have to be right about what it was before.
     */
    private fun apply(activity: Activity) {
        activity.runOnUiThread {
            if (rule.mustProtect) {
                activity.window.setFlags(
                    WindowManager.LayoutParams.FLAG_SECURE,
                    WindowManager.LayoutParams.FLAG_SECURE,
                )
            } else {
                activity.window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
            }

            Log.d(TAG, "capture protection: ${rule.mustProtect} (rule=$rule)")
        }
    }

    /** `Lemonfiber.Conceal` — a screen holding a secret has come up. */
    public class Conceal(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            rule = rule.concealing()
            apply(activity)

            return mapOf("protected" to rule.mustProtect)
        }
    }

    /** `Lemonfiber.Reveal` — that screen has gone. */
    public class Reveal(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            rule = rule.revealing()
            apply(activity)

            return mapOf("protected" to rule.mustProtect)
        }
    }

    /**
     * `Lemonfiber.IsProtected` — whether the window is protected right now.
     *
     * Answered from the rule rather than by reading the window's flags back.
     * Reading them back would be the more suspicious-sounding choice and is the
     * wrong one: `getAttributes().flags` is whatever was last set on the UI
     * thread, and this call arrives from the bridge thread, so a read can beat
     * its own write and report the previous state as the current one.
     */
    public class IsProtected(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> =
            mapOf("protected" to rule.mustProtect)
    }
}

File: bridge/resources/ios/LemonfiberFunctions.swift
import Foundation
import UIKit

/// The iOS half of lemonfiber's own native expansion.
///
/// Namespace: `Lemonfiber.*`
///
/// The decision about *when* to protect the window lives in `CaptureRule`, which
/// imports nothing from UIKit and is unit-tested. This file is what cannot be
/// tested off a device, and is kept down to the calls that need one.
///
/// **iOS has no `FLAG_SECURE`.** Android hands you one window flag that answers
/// both requirements; here they are two different mechanisms and it is worth
/// being clear which is which.
///
/// **The task-switcher snapshot** is answered by covering the window as the
/// app resigns active. The system takes its snapshot after that
/// notification, so a view added there is what ends up in the switcher. This
/// is reliable and is what every banking app does.
///
/// **Screen recording and screenshots** have no complete answer on iOS, and
/// pretending otherwise would be the dangerous thing to do. `UIScreen.isCaptured`
/// reports that a recording or mirroring is running, and this covers the window
/// while it is true, which stops a guarded screen reaching the recording. A
/// *screenshot* cannot be blocked at all; `userDidTakeScreenshotNotification`
/// arrives after the fact. So on iOS a guarded screen is protected from recording
/// and from the switcher, and a deliberate screenshot by the person holding the
/// phone is out of reach. That is a platform limit, not a gap in this file, and
/// it is written here so nobody has to rediscover it.
enum LemonfiberFunctions {
    /// What the app is showing and whether anybody is looking at it.
    ///
    /// Process-wide because the window is, and guarded by the main queue because
    /// every read and write happens there.
    nonisolated(unsafe) private static var rule: CaptureRule = .launched

    /// The view put over the window. One instance, reused.
    nonisolated(unsafe) private static var cover: UIView?

    /// Start watching the app move in and out of the foreground.
    ///
    /// Called once from `LemonfiberInit`. Without it the window is never covered,
    /// and the failure is invisible: every screen looks right, and the task
    /// switcher quietly holds the last frame.
    static func install() {
        let centre = NotificationCenter.default

        centre.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: .main
        ) { _ in
            rule = rule.backgrounded()
            apply()
        }

        centre.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { _ in
            rule = rule.foregrounded()
            apply()
        }

        // A recording can start while the app is in front of you, which is the
        // case a guarded screen is about and the one a backgrounding observer
        // never sees.
        centre.addObserver(
            forName: UIScreen.capturedDidChangeNotification,
            object: nil,
            queue: .main
        ) { _ in
            apply()
        }
    }

    /// Put the window into whatever state the rule now calls for.
    ///
    /// `isCaptured` is read here rather than folded into `CaptureRule`, because
    /// it is a fact about the device at this instant rather than about what the
    /// app is showing — and the rule is the part that is meant to be testable
    /// without one.
    private static func apply() {
        let recording = UIScreen.main.isCaptured
        let hide = rule.mustProtect || (rule.concealed && recording)

        guard
            let window = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .flatMap({ $0.windows })
                .first(where: { $0.isKeyWindow })
        else {
            return
        }

        if hide {
            let over = cover ?? makeCover()
            cover = over
            over.frame = window.bounds
            window.addSubview(over)
            window.bringSubviewToFront(over)
        } else {
            cover?.removeFromSuperview()
        }
    }

    /// The thing the task switcher gets a picture of.
    ///
    /// Deliberately opaque and deliberately empty. A blur over the real view is
    /// the prettier choice and it is a picture of the content: a heavy blur of a
    /// credential is still shaped like one, and the switcher's thumbnail is small
    /// enough that "shaped like" is most of what survives anyway.
    private static func makeCover() -> UIView {
        let over = UIView()
        over.backgroundColor = .systemBackground
        over.autoresizingMask = [.flexibleWidth, .flexibleHeight]

        return over
    }

    /// `Lemonfiber.Conceal` — a screen holding a secret has come up.
    class Conceal: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            DispatchQueue.main.async {
                rule = rule.concealing()
                apply()
            }

            return BridgeResponse.success(data: ["protected": true])
        }
    }

    /// `Lemonfiber.Reveal` — that screen has gone.
    class Reveal: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            DispatchQueue.main.async {
                rule = rule.revealing()
                apply()
            }

            return BridgeResponse.success(data: ["protected": false])
        }
    }

    /// `Lemonfiber.IsProtected` — whether the window is protected right now.
    ///
    /// Answered from the rule rather than by asking whether the cover is in the
    /// hierarchy. The cover is added on the main queue and this call arrives from
    /// the bridge, so a read of the view tree can beat its own write and report
    /// the previous state as the current one.
    class IsProtected: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            BridgeResponse.success(data: ["protected": rule.mustProtect])
        }
    }

    /// `Lemonfiber.IsInFront`: whether the app is in front of somebody right now.
    ///
    /// The rule's own record of the last resign or become-active, which is
    /// what the observers installed at launch wrote.
    class IsInFront: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            BridgeResponse.success(data: ["inFront": rule.foreground])
        }
    }
}
